### PR TITLE
Add transformation extensions

### PR DESCRIPTION
**Pull request by bdjilka**

### CODE REVIEW

File: connect/eaas/core/decorators.py [hunk 1]
```diff
@@ -121,5 +122,25 @@ def wrapper(cls):
     return wrapper
 
 
+def transformation(name, description, edit_dialog_ui):
+    def wrapper(cls):
+        if hasattr(cls, TRANSFORMATION_ATTR_NAME):
```
Comment (ffaraone) [inline connect/eaas/core/decorators.py:127]: I think this if is not needed. If people put the same decorator twice is not our problem.

**Review by Sainomori — APPROVED** [review]
👍

File: connect/eaas/core/decorators.py [hunk 2]
```diff
@@ -121,5 +122,25 @@ def wrapper(cls):
...
+                declared_transf.append({
+                    'name': name,
+                    'description': description,
+                    'edit_dialog_ui': edit_dialog_ui,
+                })
+            return cls
+        setattr(cls, TRANSFORMATION_ATTR_NAME, [{
+            'name': name,
+            'description': description,
+            'edit_dialog_ui': edit_dialog_ui,
```
Comment (ffaraone) [inline connect/eaas/core/decorators.py:139]: probably we can set here the fqn of this class like:
```
'class_fqn': f'{cls.__module__}.{cls.__name__}',
```

File: connect/eaas/core/extension.py [hunk 3]
```diff
@@ -149,3 +150,37 @@ def setup_anvil_callables(self):
                 fn = functools.partial(_invoke, value)
                 fn.__name__ = value.__name__
                 anvil.server.callable(fn)
+
+
+class TransformationBase(ApplicationBase):
```
Comment (ffaraone) [inline connect/eaas/core/extension.py:155]: No need to inherit from ApplicationBase.
The Runner will handle the set of transformation starting from the module so the way he get the extension.json descriptor will be starting from the module location.

File: connect/eaas/core/decorators.py [hunk 4]
```diff
@@ -121,5 +122,25 @@ def wrapper(cls):
...
+        if hasattr(cls, TRANSFORMATION_ATTR_NAME):
+            declared_transf = getattr(cls, TRANSFORMATION_ATTR_NAME)
+            if len(list(filter(lambda x: x['name'] == name, declared_transf))) == 0:  # noqa: B023
+                declared_transf.append({
+                    'name': name,
+                    'description': description,
+                    'edit_dialog_ui': edit_dialog_ui,
+                })
+            return cls
+        setattr(cls, TRANSFORMATION_ATTR_NAME, [{
```
Comment (ffaraone) [inline connect/eaas/core/decorators.py:136]: It should be just an object not an list of objects.
The runner will iterate over all the classes within a module (the module to which we will refer using setuptools entrypoints) that inherit from TransformationBase to build the list of all transformations

File: connect/eaas/core/extension.py [hunk 5]
```diff
@@ -149,3 +150,37 @@ def setup_anvil_callables(self):
...
+        self.input_columns = input_columns
+        self.output_columns = output_columns
+        self.stream = stream
+
+        self.transformation_settings = transformation_settings
+        self.cache = cache
+
+    @classmethod
+    def get_transformation_info(cls):
+        return getattr(cls, TRANSFORMATION_ATTR_NAME, [])
```
Comment (ffaraone) [inline connect/eaas/core/extension.py:186]: default should be None (just for validations in the future)

File: connect/eaas/core/proto.py [hunk 6]
```diff
@@ -141,6 +141,13 @@ class Schedulable(BaseModel):
     description: str
 
 
+class Transformation(BaseModel):
+    method: str
+    name: str
+    description: str
+    class_name: str
```
Comment (ffaraone) [inline connect/eaas/core/proto.py:148]: This should be `edit_dialog_ui`

File: connect/eaas/core/proto.py [hunk 7]
```diff
@@ -141,6 +141,13 @@ class Schedulable(BaseModel):
     description: str
 
 
+class Transformation(BaseModel):
+    method: str
```
Comment (ffaraone) [inline connect/eaas/core/proto.py:145]: this should be `class_fqn`

**Review by ffaraone — CHANGES_REQUESTED** [review]

**Review by ffaraone — APPROVED** [review]

**Review by d3rky — APPROVED** [review]

**Pull request merged**